app/assets/javascripts/preferences/panes/two-factor-auth/TwoFactorAuthView.tsx

```diff
@@ -0,0 +1,135 @@
+  auth: TwoFactorAuth;
+}> = observer(({ auth }) => (
+  <PreferencesGroup>
+    <PreferencesSegment>
+      <div className="flex flex-row items-center">
+        <div className="flex-grow flex flex-col">
+          <Title>Two-factor authentication</Title>
+          <Text>
+            An extra layer of security when logging in to your account.
+          </Text>
+        </div>
+        <Switch
+          checked={auth.enabled !== false}
+          onChange={() => auth.toggle2FA()}
+        />
+      </div>
+    </PreferencesSegment>
+    <PreferencesSegment>
+      {auth.enabled !== false && (
+        <TwoFactorEnabledView
+          secretKey={auth.enabled.secretKey}
+          authCode={auth.enabled.authCode}
+        />
+      )}
+
+      {auth.activation !== false && (
+        <TwoFactorActivationView activation={auth.activation} />
+      )}
+
+      {auth.enabled === false && <TwoFactorDisabledView />}
+    </PreferencesSegment>
+  </PreferencesGroup>
+));
+
+const ProgressTime: FunctionalComponent<{ time: number }> = ({ time }) => {
+  const [percent, setPercent] = useState(0);
+  const interval = time / 100;
+  useEffect(() => {
+    const tick = setInterval(() => {
+      if (percent === 100) {
+        setPercent(0);
+      } else {
+        setPercent(percent + 1);
+      }
+    }, interval);
+    return () => {
+      clearInterval(tick);
+    };
+  });
+  return <CircleProgress percent={percent} />;
+};
+
+const TwoFactorEnabledView: FunctionComponent<{
+  secretKey: string;
+  authCode: string;
+}> = ({ secretKey, authCode }) => {
+  const download = (
+    <IconButton
+      icon="download"
+      onClick={() => {
+        downloadSecretKey(secretKey);
+      }}
+    />
+  );
+  const copy = (
+    <IconButton
+      icon="copy"
+      onClick={() => {
+        navigator?.clipboard?.writeText(secretKey);
+      }}
+    />
+  );
+  const progress = <ProgressTime time={30000} />;
+  return (
+    <div className="flex flex-row gap-4">
+      <div className="flex-grow flex flex-col">
+        <Text>Secret Key</Text>
+        <DecoratedInput
+          disabled={true}
+          right={[copy, download]}
+          text={secretKey}
+        />
+      </div>
+      <div className="w-30 flex flex-col">
+        <Text>Authentication Code</Text>
+        <DecoratedInput disabled={true} text={authCode} right={[progress]} />
+      </div>
+    </div>
+  );
+};
+
+const TwoFactorDisabledView: FunctionComponent = () => (
+  <Text>
+    Enabling two-factor authentication will sign you out of all other sessions.{' '}
+    <a
+      target="_blank"
+      href="https://standardnotes.com/help/21/where-should-i-store-my-two-factor-authentication-secret-key"
+    >
+      Learn more
+    </a>
+  </Text>
+);
+
+export const TwoFactorActivationView: FunctionComponent<{
+  activation: TwoFactorActivation;
+}> = observer(({ activation: act }) => (
+  <>
+    {act.step === 'scan-qr-code' && <ScanQRCode activation={act} />}
```

Comment:
Perhaps we could implement this with a switch?

Reply:
It's the same effect, but this way felt more natural since we're developing a template. It's more flexible to manipulation, e.g. we can add an extra condition, we can add decorations around it, if needed, stuff like that. I'd leave it as is if it's all the same to you.